src/use_pathspec.md

```diff
@@ -0,0 +1,479 @@
+
+/**
+ * example Pegasus schema of a record containing record field
+ */
+record RecordTest {
+  recordField: RecordBar
+}
+```
+And that record expands to
+```pdl
+namespace com.linkedin.pegasus.examples
+
+record RecordBar {
+  location: string
+}
+```
+
+Then the PathSpec reference to the nested field "location" is
+```
+/recordField/location
+```
+### Map and Array type fields
+Map and Array are two collection type used in Pagasus. 
+
+##### Map type
+The map has a key followed by the value referenced. Recall PathSpec's semantic is a reference to a component within a complex data object. The map entries were referenced by their keys. 
+
+For example, here is a example PathSpec that can be used to traverse the map.
+
+```
+/mapField/<AKeyInMap>
+```
+Note that the above PathSpec is referring to the component key-ed by the key in the map. If the value is of a primitive type, the above PathSpec can be implicitly used to refer to the primitive type value.
+
+If you want to specifically refer to all the key fields, the syntax is to use the keyword "$key"
+```
+/mapField/$key
+```
+
+if the value held in the map is a record type and you want to refer to that record, the PathSpec could be
+
+```
+/mapField/<AKeyInMap>/<someRecordField>
+```
+
+Another common use case is that you want to use wildcard (represented by symbol "*" ) to "select all keys", if the interested path is to a field for all entries in this map, then it would be
+
+```
+/mapField/*/<recordField>
+```
+
+It is worth noting that in Pegasus, the map keys are always of string types.
+
+Here is an concrete example of a record, containing a map field, and that map field's map has value of record type. The example also defines a "recordInlineMap" field for similar demenstration purpose
+```pdl
+namespace com.linkedin.pegasus.examples
+
+/**
+ * a record containing a map field, which holds values of record type
+ */
+record RecordWithMapFields {
+
+  recordMap: map[string, RecordBar]
+
+  recordInlineMap: map[string, record RecordInMap {
+    f: int
+  }]
+}
+```
+The PathSpec to refer to the record field, and the inline map field, are
+```
+/recordMap/*/location
+/recordInlineMap/*/f
+```
+
+##### Array type
+Array type is somewhat similar to Map type in the sense that the length might not be definite. For example it could be defined as such
+```pdl
+namespace com.linkedin.pegasus.examples
+
+/**
+ * a record containing an array field, which holds values of record type
+ */
+record RecordWithArrayFields {
+
+  recordArray: array[RecordBar]
+
+  recordInlineArray: array[record RecordInArray {
+    f: int
+  }]
+}
+```
+Then similarly if you are interested in some fields only, you can use wildCard to select all elements so you can refer to a field
+```
+/recordArray/*/location
+/recordInlineArray/*/f
+```
+
+One big difference between Array and Map is that in Array, we support reference to the range. It is achieved by specifying attributes aforementioned. 1
```

Comment:
It is achieved by using the 'start' and 'count' attributes.

(Hard to remember which attributes you are referring to, or if someone reads this section directly)